File: backend/routers/whatsapp.py
"""
WhatsApp router — Send messages and receive Twilio webhooks
"""
from fastapi_cache.decorator import cache
from fastapi import APIRouter, Request, HTTPException, Form, Response
from datetime import datetime
from models.message import SendMessageRequest
from services.firebase_service import get_db
from services.twilio_service import send_whatsapp_message, send_bulk_messages
from auth import get_current_user
from fastapi import Depends

router = APIRouter(prefix="/whatsapp", tags=["whatsapp"])


@router.post("/send")
async def send_message(req: SendMessageRequest, user: dict = Depends(get_current_user)):
    """
    Send a single WhatsApp message via Twilio.
    Saves the attempt to the local DB regardless of outcome.
    """
    db = get_db()

    result = send_whatsapp_message(req.to, req.body, req.media_url)

    
    msg_data = {
        "userId": user["uid"],
        "leadId": req.lead_id or "",
        "phone": req.to,
        "direction": "outbound",
        "body": req.body,
        "status": "sent" if result["success"] else "failed",
        "twilioSid": result.get("sid"),
        "createdAt": datetime.utcnow().isoformat(),
    }
    try:
        db.collection("messages").add(msg_data)
    except Exception:
        pass

    if not result["success"]:
        twilio_code = result.get("code")
        raw_error  = result.get("error", "Failed to send message")

        
        TWILIO_ERRORS = {
            63007: (
                "Sandbox opt-in required: the recipient must first send "
                "'join <your-sandbox-word>' to +14155238886 on WhatsApp before "
                "messages can be delivered."
            ),
            63031: (
                "Cannot send to yourself: the To and From numbers must be different. "
                "Your number (+917996633015) joined the sandbox but cannot receive messages "
                "sent FROM the same account that owns it. Try sending to a different WhatsApp number."
            ),
            63016: "The recipient's WhatsApp number is invalid or not registered on WhatsApp.",
            21211: "Invalid 'To' phone number — include the country code, e.g. +91XXXXXXXXXX.",
            21608: "The 'To' number is not a valid mobile number.",
            21614: "Phone number is not a valid mobile number.",
            20003: "Twilio authentication failed — check TWILIO_ACCOUNT_SID and TWILIO_AUTH_TOKEN in .env.",
        }

        friendly = TWILIO_ERRORS.get(twilio_code, raw_error)
        pass
        raise HTTPException(
            status_code=400,
            detail={"message": friendly, "twilio_code": twilio_code, "raw": raw_error},
        )

    return {"success": True, "sid": result.get("sid"), "status": result.get("status")}



@router.post("/broadcast")
async def broadcast_message(campaign_id: str, message: str, phone_numbers: list[str], user: dict = Depends(get_current_user)):
    """
    Send a message to multiple numbers (campaign broadcast).
    """
    db = get_db()
    results = send_bulk_messages(phone_numbers, message)
    
    
    campaign_ref = db.collection("campaigns").document(campaign_id)
    campaign_ref.update({
        "sentCount": results["sent"],
        "failedCount": results["failed"],
        "status": "completed",
        "completedAt": datetime.utcnow().isoformat(),
    })
    
    return results


@router.post("/webhook")
async def twilio_webhook(
    request: Request,
    From: str = Form(None),
    Body: str = Form(None),
    MessageSid: str = Form(None),
    MessageStatus: str = Form(None),
    To: str = Form(None),
):
    """
    Twilio webhook for incoming WhatsApp messages and status updates.
    Configure this URL in Twilio Console → Messaging → WhatsApp Sandbox.
    """
    db = get_db()
    
    if Body and From:
        
        pass
        
        
        phone_cleaned = From.replace("whatsapp:", "").replace("+91", "")
        leads = db.collection("leads").where("phone", ">=", phone_cleaned).limit(1).stream()
        lead_id = ""
        for lead in leads:
            lead_id = lead.id
            break
        
        
        db.collection("messages").add({
            "leadId": lead_id,
            "phone": From.replace("whatsapp:", ""),
            "direction": "inbound",
            "body": Body,
            "status": "delivered",
            "twilioSid": MessageSid,
            "createdAt": datetime.utcnow().isoformat(),
        })
    
    elif MessageStatus and MessageSid:
        
        pass
        msgs_ref = db.collection("messages").where("twilioSid", "==", MessageSid).stream()
        for msg_doc in msgs_ref:
            try:
                
                msg_doc.reference.update({"status": MessageStatus})
            except AttributeError:
                
                db.collection("messages").document(msg_doc.id).update({"status": MessageStatus})
    
    
    return Response(content='<?xml version="1.0" encoding="UTF-8"?><Response></Response>', media_type="application/xml")


@router.get("/messages")
@cache(expire=30)
async def get_messages(lead_id: str = None, user: dict = Depends(get_current_user)):
    """Get messages for a user, optionally filtered by lead."""
    db = get_db()
    query = db.collection("messages").where("userId", "==", user["uid"])
    if lead_id:
        query = query.where("leadId", "==", lead_id)
        
    msgs = query.stream()
    result = [{"id": m.id, **m.to_dict()} for m in msgs]
    
    result.sort(key=lambda x: x.get("createdAt", ""), reverse=True)
    return result

File: backend/src/routes/whatsapp.js
// =============================================
// WhatsApp Routes — Twilio Integration
// POST /api/whatsapp/send         → Send a WhatsApp message
// POST /api/whatsapp/webhook      → Twilio status callback
// GET  /api/whatsapp/messages     → Get message history
// POST /api/whatsapp/bulk         → Bulk send campaign messages
// =============================================
const express = require('express')
const router = express.Router()
const twilio = require('twilio')
const { db } = require('../firebase')
const { verifyToken } = require('../middleware/auth')

// Initialize Twilio client
function getTwilioClient() {
  const sid = process.env.TWILIO_ACCOUNT_SID
  const token = process.env.TWILIO_AUTH_TOKEN
  if (!sid || !token || sid.startsWith('AC') === false) {
    console.warn('⚠️  Twilio not configured. Set TWILIO_ACCOUNT_SID and TWILIO_AUTH_TOKEN in .env')
    return null
  }
  return twilio(sid, token)
}

// ─────────────────────────────────────────────
// POST /api/whatsapp/send
// Send a single WhatsApp message via Twilio
// ─────────────────────────────────────────────
router.post('/send', verifyToken, async (req, res) => {
  const { phone, message } = req.body

  if (!phone || !message) {
    return res.status(400).json({ error: 'phone and message are required' })
  }

  const client = getTwilioClient()
  
  if (!client) {
    // Demo mode — simulate success
    console.log(`[DEMO] Would send WhatsApp to ${phone}: "${message}"`)
    
    // Save to Firestore if available
    if (db) {
      await db.collection('messages').add({
        phone,
        body: message,
        direction: 'outbound',
        status: 'sent',
        twilioSid: `DEMO_${Date.now()}`,
        userId: req.user.uid,
        createdAt: new Date().toISOString(),
      })
    }

    return res.json({
      success: true,
      sid: `DEMO_${Date.now()}`,
      status: 'sent',
      message: 'Sent in demo mode (Twilio not configured)',
    })
  }

  try {
    // Normalize phone number to WhatsApp format
    let toNumber = phone.replace(/\s/g, '').replace(/[^\d+]/g, '')
    if (!toNumber.startsWith('+')) toNumber = '+' + toNumber
    const to = `whatsapp:${toNumber}`
    const from = process.env.TWILIO_WHATSAPP_NUMBER || 'whatsapp:+14155238886'

    const msg = await client.messages.create({
      from,
      to,
      body: message,
    })

    console.log(`✅ WhatsApp sent to ${to} | SID: ${msg.sid}`)

    // Log to Firestore
    if (db) {
      await db.collection('messages').add({
        phone: toNumber,
        body: message,
        direction: 'outbound',
        status: msg.status,
        twilioSid: msg.sid,
        userId: req.user.uid,
        createdAt: new Date().toISOString(),
      })
    }

    res.json({
      success: true,
      sid: msg.sid,
      status: msg.status,
      to: msg.to,
    })
  } catch (err) {
    console.error('❌ Twilio Error:', err.message)
    res.status(500).json({
      error: err.message,
      code: err.code,
    })
  }
})

// ─────────────────────────────────────────────
// POST /api/whatsapp/bulk
// Send messages to multiple leads (campaign)
// ─────────────────────────────────────────────
router.post('/bulk', verifyToken, async (req, res) => {
  const { phones, message, campaignId } = req.body

  if (!phones || !Array.isArray(phones) || phones.length === 0) {
    return res.status(400).json({ error: 'phones array is required' })
  }
  if (!message) {
    return res.status(400).json({ error: 'message is required' })
  }

  const client = getTwilioClient()
  const results = []
  let sentCount = 0
  let failedCount = 0

  for (const phone of phones) {
    try {
      let toNumber = phone.replace(/\s/g, '').replace(/[^\d+]/g, '')
      if (!toNumber.startsWith('+')) toNumber = '+' + toNumber
      const to = `whatsapp:${toNumber}`
      const from = process.env.TWILIO_WHATSAPP_NUMBER || 'whatsapp:+14155238886'

      let sid, status
      if (client) {
        const msg = await client.messages.create({ from, to, body: message })
        sid = msg.sid
        status = msg.status
      } else {
        sid = `DEMO_${Date.now()}_${Math.random().toString(36).substr(2, 6)}`
        status = 'sent'
      }

      // Log each message
      if (db) {
        await db.collection('messages').add({
          phone: toNumber,
          body: message,
          direction: 'outbound',
          status,
          twilioSid: sid,
          campaignId: campaignId || null,
          userId: req.user.uid,
          createdAt: new Date().toISOString(),
        })
      }

      sentCount++
      results.push({ phone: toNumber, sid, status, success: true })

      // Small delay to avoid rate limiting
      await new Promise(r => setTimeout(r, 100))
    } catch (err) {
      failedCount++
      results.push({ phone, error: err.message, success: false })
    }
  }

  // Update campaign stats if provided
  if (campaignId && db) {
    await db.collection('campaigns').doc(campaignId).update({
      sentCount: sentCount,
      failedCount: failedCount,
      status: 'completed',
      completedAt: new Date().toISOString(),
    })
  }

  res.json({
    success: true,
    total: phones.length,
    sent: sentCount,
    failed: failedCount,
    results,
  })
})

// ─────────────────────────────────────────────
// POST /api/whatsapp/webhook
// Twilio Status Callback Webhook
// ─────────────────────────────────────────────
router.post('/webhook', express.urlencoded({ extended: false }), async (req, res) => {
  const { MessageSid, MessageStatus, To, From, Body } = req.body

  console.log(`📱 Twilio Webhook: ${MessageStatus} | ${MessageSid}`)

  // Update message status in Firestore
  if (db && MessageSid) {
    try {
      const snap = await db.collection('messages')
        .where('twilioSid', '==', MessageSid)
        .limit(1)
        .get()

      if (!snap.empty) {
        await snap.docs[0].ref.update({ status: MessageStatus })
      }

      // If it's an INBOUND message, save it
      if (!MessageSid.startsWith('SM') && Body) {
        await db.collection('messages').add({
          phone: From?.replace('whatsapp:', ''),
          body: Body,
          direction: 'inbound',
          status: 'received',
          twilioSid: MessageSid,
          createdAt: new Date().toISOString(),
        })
      }
    } catch (err) {
      console.error('Error updating message status:', err.message)
    }
  }

  // Twilio expects a 200 TwiML response
  res.type('text/xml')
  res.send(`<?xml version="1.0" encoding="UTF-8"?><Response></Response>`)
})

// ─────────────────────────────────────────────
// GET /api/whatsapp/messages
// Get recent message history for the user
// ─────────────────────────────────────────────
router.get('/messages', verifyToken, async (req, res) => {
  if (!db) {
    return res.json({ messages: [] })
  }

  try {
    const snap = await db.collection('messages')
      .where('userId', '==', req.user.uid)
      .orderBy('createdAt', 'desc')
      .limit(100)
      .get()

    const messages = snap.docs.map(d => ({ id: d.id, ...d.data() }))
    res.json({ messages, count: messages.length })
  } catch (err) {
    console.error('Error fetching messages:', err)
    res.status(500).json({ error: err.message })
  }
})

module.exports = router
